resourcemanager/gpool/spool/spool.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spool
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/tidb/metrics"
+	"github.com/pingcap/tidb/resourcemanager"
+	"github.com/pingcap/tidb/resourcemanager/gpool"
+	"github.com/pingcap/tidb/resourcemanager/pooltask"
+	"github.com/pingcap/tidb/resourcemanager/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/mathutil"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+)
+
+// Pool is a single goroutine pool. it can not reuse the goroutine.
+type Pool struct {
+	wg                 sync.WaitGroup
+	mu                 sync.Mutex
+	options            *Options
+	capacity           atomic.Int32
+	running            atomic.Int32
+	isStop             atomic.Bool
+	concurrencyMetrics prometheus.Gauge
+	taskManager        pooltask.TaskManager[any, any, any, any, pooltask.NilContext]
+	gpool.BasePool
+}
+
+// NewPool create a single goroutine pool.
+func NewPool(name string, size int32, component util.Component, options ...Option) (*Pool, error) {
+	opts := loadOptions(options...)
+	result := &Pool{
+		options:            opts,
+		concurrencyMetrics: metrics.PoolConcurrencyCounter.WithLabelValues(name),
+		taskManager:        pooltask.NewTaskManager[any, any, any, any, pooltask.NilContext](size), // TODO: this general type
+	}
+	result.SetName(name)
+	result.capacity.Store(size)
```

Comment:
Do we need to check the validity of `size` here?

Reply:
Yes, I added.